impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -1283,14 +1386,70 @@ private static long safeExpireTime(long now, Duration duration) {
     return result;
   }
   
-  private void enforceCapacity(int delta) {
-    for (int attempts = 0, evicted = 0; attempts < ATTEMPT_RATIO * delta && evicted < EVICTION_RATIO * delta
-            && capacity < map.size(); attempts++) {
+  private void enforceByteCapacity() {
+    while (capacity < currentUsageinBytes.get()) {
+      evict();
+    }
+  }
+ 
+  private void incrementCurrentUsageInBytesIfRequired(long delta) {
+    if(byteSized) {
+      while(true) {
+        long expect =  currentUsageinBytes.get();
+        long update = expect + delta;
+        if(currentUsageinBytes.compareAndSet(expect, update)) {
+          break;
+        }
+      }
+    }
+  }
+  
+  private void decrementCurrentUsageInBytesIfRequired(long delta) {
+    if(byteSized) {
+      while(true) {
+        long expect =  currentUsageinBytes.get();
+        long update = expect - delta;
+        if(update < 0l) {
+          //Either this happened due to race 
+          //or the size estimation screwed up somewhere
+          //come out silently or throw ?
+          throw new AssertionError("It's screwed");
+        }
+        if(currentUsageinBytes.compareAndSet(expect, update)) {
+          break;
+        }
+      }
+    }
+  }
+  
+  protected long getCurrentUsageInBytes() {
+    if (byteSized) {
+      return currentUsageinBytes.get();
+    }
+    return 0l;
+  }
+  
+  protected void enforceCapacity(long delta) {
+    if(byteSized) {
```

Comment:
This is wrong. We need to keep the logic of eviction doing bounded work, even if it means that the cache stays over size for a while. Here you have no bounds on the number of elements to evict. Also, once this logic is applied, then capacity enforcement might need to happen not only after the size grows.